admin/menu/class-base-menu.php

```diff
@@ -0,0 +1,192 @@
+<?php
+/**
+ * WPSEO plugin file.
+ *
+ * @package WPSEO\Admin\Menu
+ */
+
+/**
+ * Admin menu base class.
+ */
+abstract class WPSEO_Base_Menu implements WPSEO_WordPress_Integration {
+
+	/** @var WPSEO_Menu Menu */
+	protected $menu;
+
+	/**
+	 * Constructs the Admin Menu.
+	 *
+	 * @param WPSEO_Menu $menu Menu to use.
+	 */
+	public function __construct( WPSEO_Menu $menu ) {
+		$this->menu = $menu;
+	}
+
+	/**
+	 * Registers all hooks to WordPress.
+	 *
+	 * @return void
+	 */
+	public abstract function register_hooks();
+
+	/**
+	 * Returns the list of registered submenu pages.
+	 *
+	 * @return array List of registered submenu pages.
+	 */
+	public abstract function get_submenu_pages();
+
+	/**
+	 * Creates a submenu formatted array.
+	 *
+	 * @param string     $page_title Page title to use.
+	 * @param string     $page_slug  Page slug to use.
+	 * @param callable   $callback   Optional. Callback which handles the page request.
+	 * @param callable[] $hook       Optional. Hook to trigger when the page is registered.
+	 *
+	 * @return array Formatted submenu.
+	 */
+	protected function get_submenu_page( $page_title, $page_slug, $callback = null, $hook = null ) {
+		if ( $callback === null ) {
+			$callback = $this->get_admin_page_callback();
+		}
+
+		return array(
+			$this->get_page_identifier(),
+			'',
+			$page_title,
+			$this->get_manage_capability(),
+			$page_slug,
+			$callback,
+			$hook,
+		);
+	}
+
+	/**
+	 * Registers submenu pages as menu pages.
+	 *
+	 * @param array $submenu_pages List of submenu pages.
+	 */
+	protected function register_menu_pages( $submenu_pages ) {
+		if ( ! is_array( $submenu_pages ) || $submenu_pages === array() ) {
+			return;
+		}
+
+		// Loop through submenu pages and add them.
+		foreach ( $submenu_pages as $submenu_page ) {
+			if ( $submenu_page[3] === $this->get_manage_capability() ) {
+				continue;
+			}
+
+			// Register submenu as menu page.
+			add_menu_page(
+				'Yoast SEO: ' . $submenu_page[2],
+				$submenu_page[2],
+				$submenu_page[3],
+				$submenu_page[4],
+				$submenu_page[5],
+				WPSEO_Utils::get_icon_svg(),
+				'99.31337'
+			);
+		}
+	}
+
+	/**
+	 * Registers the submenu pages.
+	 *
+	 * This is only done when the user has the capability to manage options,
+	 * thus all capabilities can be set to this capability.
+	 *
+	 * @param array $submenu_pages List of submenu pages to register.
+	 *
+	 * @return void
+	 */
+	protected function register_submenu_pages( $submenu_pages ) {
```

Comment:
This method is pretty complex. Is it possible to reduce the complexity?

Reply:
I used `array_walk()` for this as well and implemented `register_submenu_page()` used by that.